File: ruby/gilded_rose.rb
class GildedRose
  attr_reader :items

  def initialize(items)
    @items = items
  end

  def update_quality
    items.each { update_item(_1) }
  end

  def update_item(item)
    case item.name
    when /Aged Brie/
      item.sell_in -= 1

      item.quality += case item.sell_in
                      when ...0 then 2
                      else           1
                      end

      item.quality  = item.quality.clamp(0, 50)
    when /Backstage pass/
      item.sell_in -= 1

      item.quality += case item.sell_in
                      when ...0  then -item.quality
                      when ...5  then 3
                      when ...10 then 2
                      else            1
                      end

      item.quality  = item.quality.clamp(0, 50)
    when /Sulfuras/
      # no-op
    else
      item.sell_in -= 1

      item.quality += case item.sell_in
                      when ...0 then -2
                      else           -1
                      end

      item.quality = item.quality.clamp(0, 50)
    end
  end
end

File: js/gilded_rose.js
export class Shop {
  constructor(items = []) {
    this.items = items;
  }

  updateQuality() {
    this.items.forEach(this.updateItem);
    return this.items;
  }

  updateItem(item) {
    switch (item.name) {
      case "Aged Brie":
        item.sellIn--;
        item.quality += item.sellIn < 0 ? 2 : 1;
        if (item.quality > 50) item.quality = 50;
        break;
      case "Backstage passes to a TAFKAL80ETC concert":
        item.sellIn--;
        if (item.sellIn >= 10) {
          item.quality += 1;
        } else if (item.sellIn >= 5) {
          item.quality += 2;
        } else if (item.sellIn >= 0) {
          item.quality += 3;
        } else {
          item.quality = 0;
        }
        if (item.quality > 50) item.quality = 50;
        break;
      case "Sulfuras, Hand of Ragnaros":
        break;
      default:
        item.sellIn--;
        item.quality -= item.sellIn < 0 ? 2 : 1;
        if (item.quality < 0) item.quality = 0;
    }
  }
}
